### PR TITLE
JWT Auth Support 

### PR DESCRIPTION
**Pull request by voicenter**
adding auth parameter to UA for supporting :
https://opensips.org/docs/modules/3.1.x/auth_jwt.html

### CODE REVIEW

**Comment by ibc** [comment]
Not sure if that commit is the last one. Did you actually really run `gulp lint`?

```
ibc at ibc-macbook-pro in /tmp/JsSIP on master
$ gulp lint
[11:18:42] Using gulpfile /private/tmp/JsSIP/gulpfile.js
[11:18:42] Starting 'lint'...
[11:18:43]
/private/tmp/JsSIP/lib/RTCSession.js
  282:1    error  Expected indentation of 4 spaces but found 5                           indent
  282:6    error  Expected space(s) after "if"                                           keyword-spacing
  282:38   error  Multiple spaces found before '&&'                                      no-multi-spaces
  282:60   error  Expected parentheses around arrow function argument                    arrow-parens
  282:98   error  Strings must use singlequote                                           quotes
  282:134  error  Opening curly brace appears on the same line as controlling statement  brace-style
  282:134  error  Missing space before opening brace                                     space-before-blocks
  283:25   error  Unexpected string concatenation                                        prefer-template
  283:25   error  Strings must use singlequote                                           quotes
  283:78   error  There should be no spaces inside this paren                            space-in-parens
  283:79   error  Missing semicolon                                                      semi

/private/tmp/JsSIP/lib/Registrator.js
  43:1    error  Expected indentation of 4 spaces but found 1 tab  indent
  44:1    error  Expected indentation of 4 spaces but found 1 tab  indent
  44:2    error  Expected space(s) after "if"                      keyword-spacing
  44:29   error  Multiple spaces found before 'this'               no-multi-spaces
  44:53   error  Unexpected string concatenation                   prefer-template
  44:100  error  Missing semicolon                                 semi

✖ 17 problems (17 errors, 0 warnings)
  17 errors and 0 warnings potentially fixable with the `--fix` option.
```

**Comment by ibc** [comment]
Please, do really run `gulp lint` in your fork.

**Comment by voicenter** [comment]
Oops sorry missed pushing those files, many thanks...
let me know if I should prepare any documentation about how to use it.
or just to add the auth to the parameters in the UA properties.
many thanks in advanced
Shlomi

File: lib/RTCSession.js [hunk 1]
```diff
@@ -278,7 +278,11 @@ module.exports = class RTCSession extends EventEmitter
     this._rtcAnswerConstraints = options.rtcAnswerConstraints || null;
 
     this._data = options.data || this._data;
-
+    // Check Authorization: Bearer
```
Comment (ibc) [inline lib/RTCSession.js]:
```suggestion
    // Check Authorization: Bearer.
```

File: lib/RTCSession.js [hunk 2]
```diff
@@ -278,7 +278,11 @@ module.exports = class RTCSession extends EventEmitter
     this._rtcAnswerConstraints = options.rtcAnswerConstraints || null;
 
     this._data = options.data || this._data;
-
+    // Check Authorization: Bearer.
+    if (this._ua.configuration.auth && extraHeaders.filter((extraHeader) => extraHeader.startsWith('Authorization: Bearer')).length==0)
```
Comment (ibc) [inline lib/RTCSession.js]: This is not ok:

* We must not mandate SIP headers in so strict low/upper case.
* Anyway, it looks like in order to enable JWT auth, the user must call `ua.call()` with "Authorization: Bearer" in `options.extraHeaders`.  No way, this is a hack. Let's create a different `option.useJwtAuth: boolean` instead.

**Comment by jmillan** [comment]
> A proper new option (something such as enableJwtAuth) in options given to ua.call().
It completely lacks JWT auth support for any other type of SIP request being sent.
So, this should not be done in ua.call() at all. Instead, JWT auth should be added (if enabled via settings) for all sent SIP requests that require authentication.
@jmillan your thoughts?

Definitely such an option is needed. Also [RequestSender](https://github.com/versatica/JsSIP/blob/master/lib/RequestSender.js) should be modified, considering such new option. It's exactly the place where the header should be handled, not in `RTCSession`

**Comment by jmillan** [comment]
@voicenter, let us know in case you need some guidance.

**Comment by voicenter** [comment]
Hi ... 
many thanks for the feedback,
hope that we understand you well, 
we didn't want to be too aggressive and pushing it anywhere,
but actually we totally agree  about it 👍  
let me know if you want us to rename the [auth parameter](https://github.com/voicenter/JsSIP/blob/master/lib/Config.js#L126) to something less generic,
like auth_token or jwt_token , let me know what you prefer. 
Thanks again ... 
Shlomi

File: lib/RequestSender.js [hunk 3]
```diff
@@ -68,6 +68,11 @@ module.exports = class RequestSender
         this.clientTransaction = new Transactions.NonInviteClientTransaction(
           this._ua, this._ua.transport, this._request, eventHandlers);
     }
+    // Check if auth token is set in the configuration and no other Authorization header is set before
+    if (this._ua._configuration.auth && !this._request.headers.Authorization)
+    {
+      this._request.headers.Authorization = [ `Bearer ${this._ua._configuration.auth}` ];
+    }
```
Comment (jmillan) [inline lib/RequestSender.js]: When using JWT authorization we are not acting upon server side 401/407 responses, so it is fine to just add the JWT if exists.

```suggestion
    // If authorization JWT is present, use it.
    if (this._ua._configuration.authorization_jwt)
    {
      this._request.headers.Authorization = [ `Bearer ${this._ua._configuration.auth}` ];
    }
```

**Comment by jmillan** [comment]
Lets call the configuration option `authorization_jwt` pleas

**Comment by jmillan** [comment]
New JsSIP version has been released with the new configuration parameter.

Thanks @voicenter :-)

File: lib/RequestSender.js [hunk 4]
```diff
@@ -68,6 +68,11 @@ module.exports = class RequestSender
         this.clientTransaction = new Transactions.NonInviteClientTransaction(
           this._ua, this._ua.transport, this._request, eventHandlers);
     }
+    // If authorization JWT is present, use it.
+    if (this._ua._configuration.authorization_jwt)
+    {
+      this._request.headers.Authorization = [ `Bearer ${this._ua._configuration.authorization_jwt}` ];
```
Comment (ibc) [inline lib/RequestSender.js]: Is this ok? It's assigning an array as header value...

**Comment by ibc** [comment]
Thanks @voicenter.

Just wondering if this is ok: https://github.com/versatica/JsSIP/pull/610/files#r399869207

**Pull request merged**